spicyc generates uncompilable linker glue for list fields

For the following module

```.ruby
# foo.spicy
module foo;

public type A = unit {
    bs: B[1] {}
};

type B = unit {};
```

`spicyc` currently generates linker glue which cannot be compiled since not all needed types are emitted,

```
$ spicyc -l foo.spicy -o foo.cc
$ $(spicy-config --cxx --cxxflags) foo.cc
pe.cc:15:96: error: no member named 'B' in namespace '__hlt::foo'
    extern void __hook_A___on_bs(const hilti::rt::Vector<hilti::rt::ValueReference<__hlt::foo::B>>& __dd, hilti::rt::ValueReference<A>& __self);
                                                                                   ~~~~~~~~~~~~^
pe.cc:16:101: error: no member named 'B' in namespace '__hlt::foo'
    extern void __hook_A___on_bs_137e(const hilti::rt::Vector<hilti::rt::ValueReference<__hlt::foo::B>>& __dd, hilti::rt::ValueReference<A>& __self);
                                                                                        ~~~~~~~~~~~~^
pe.cc:22:104: error: no member named 'B' in namespace '__hlt::foo'
extern void __hlt::foo::__hook_A___on_bs(const hilti::rt::Vector<hilti::rt::ValueReference<__hlt::foo::B>>& __dd, hilti::rt::ValueReference<A>& __self) { __hlt::foo::__hook_A___on_bs_137e(__dd, __self); }
                                                                                           ~~~~~~~~~~~~^
3 errors generated.
```

If the field `bs` is instead declared as `B` instead of `B[1]` the generated code can be compiled, so this seems to be due to the way we compute the list of needed types.

Simply reordering the unit declarations or declaring `B` as `public` in the Spicy file does not make the issue go away, so I am unsure what a workaround could be.

Above reproducer was reduced from https://github.com/zeek/spicy-analyzers/blob/main/analyzer/file/pe/pe.spicy.

--- a/hilti/toolchain/src/compiler/codegen/codegen.cc
+++ b/hilti/toolchain/src/compiler/codegen/codegen.cc
@@ -7,6 +7,7 @@
 #include <hilti/ast/id.h>
 #include <hilti/ast/module.h>
 #include <hilti/ast/scope-lookup.h>
+#include <hilti/ast/type.h>
 #include <hilti/ast/types/struct.h>
 #include <hilti/base/logger.h>
 #include <hilti/base/util.h>
@@ -288,6 +289,9 @@ struct Visitor : hilti::visitor::PreOrder<void, Visitor> {
 
             for ( const auto& p : ft.parameters() ) {
                 auto type = p.type();
+
+                if ( type::isIterable(type) )
+                    type = type.elementType();
 
                 while ( type::isReferenceType(type) )
                     type = type.dereferencedType();

--- a/tests/spicy/types/unit/vector-field-link-glue.spicy
+++ b/tests/spicy/types/unit/vector-field-link-glue.spicy
@@ -1,0 +1,14 @@
+# @TEST-EXEC: spicyc -l %INPUT -o foo.cc
+# @TEST-EXEC: $(spicy-config --cxx --cxxflags --debug) -c foo.cc
+#
+# @TEST-DOC: Checks that we emit type defs in the linker glue for vector fields.
+
+# This is a regression test for #966.
+
+module foo;
+
+public type A = unit {
+    bs: B[1] {}
+};
+
+type B = unit {};

